src/upgrades/machine_upgrade.py

```diff
@@ -123,12 +123,28 @@ def authorized(self) -> bool:
         """Whether this unit is authorized to upgrade.
 
         Only applies to machine charm.
+
+        Raises:
+            PrecheckFailed: App is not ready to upgrade
         """
         assert self._unit_workload_container_version != self._app_workload_container_version
+        assert self.versions_set
         for index, unit in enumerate(self._sorted_units):
             if unit.name == self._unit.name:
                 # Higher number units have already upgraded
-                if index == 1:
+                if index == 0:
+                    if (
+                        json.loads(self._app_databag["versions"])["charm"]
+                        == self._current_versions["charm"]
+                    ):
+                        # Assumes charm version uniquely identifies charm revision
+                        logger.debug("Rollback detected. Skipping pre-upgrade check")
```

Comment:
Ah great catch, thanks for that. I blindly copy pasta'd 🍝 It should always be ran even in a rollback. Would you mind checking that my fix here is correct?

Reply:
> the pre-upgrade-check is skipped during rollback (because you're probably rolling back because something went wrong). It is not skipped for downgrades

Ah okay, thank you. This makes me feel much better. Thank you Carl. Also thanks for this thorough description!

Reply:
I believe I reverted the change. Is there something else that needs to be reverted?